File: test-aicf-efficiency.js
#!/usr/bin/env node

/**
 * AICF Token Efficiency Testing
 * Measures actual compression rates and semantic preservation
 */

const fs = require('fs');
const path = require('path');

function countTokens(text) {
  // Approximate GPT-style tokenization
  return text.replace(/\s+/g, ' ').split(/[\s\W]+/).filter(Boolean).length;
}

function analyzeFile(filePath) {
  if (!fs.existsSync(filePath)) {
    console.log(`❌ File not found: ${filePath}`);
    return null;
  }

  const content = fs.readFileSync(filePath, 'utf8');
  const lines = content.split('\n').filter(Boolean);
  const tokens = countTokens(content);
  const size = Buffer.byteLength(content, 'utf8');

  return {
    path: path.basename(filePath),
    lines: lines.length,
    tokens,
    size,
    compression_ratio: tokens / size,
    semantic_blocks: (content.match(/@[A-Z_]+/g) || []).length
  };
}

function main() {
  console.log('🧪 AICF Token Efficiency Analysis\n');
  
  const aicfDir = './.aicf';
  const files = ['index.aicf', 'conversations.aicf', 'decisions.aicf', 'work-state.aicf', 'technical-context.aicf'];
  
  let totalTokens = 0;
  let totalSize = 0;
  let totalBlocks = 0;
  
  files.forEach(file => {
    const analysis = analyzeFile(path.join(aicfDir, file));
    if (analysis) {
      console.log(`📊 ${analysis.path}:`);
      console.log(`   Lines: ${analysis.lines}`);
      console.log(`   Tokens: ${analysis.tokens}`);
      console.log(`   Size: ${analysis.size} bytes`);
      console.log(`   Semantic Blocks: ${analysis.semantic_blocks}`);
      console.log(`   Token/Byte Ratio: ${analysis.compression_ratio.toFixed(4)}`);
      console.log('');
      
      totalTokens += analysis.tokens;
      totalSize += analysis.size;
      totalBlocks += analysis.semantic_blocks;
    }
  });
  
  console.log('📈 OVERALL STATISTICS:');
  console.log(`   Total Tokens: ${totalTokens.toLocaleString()}`);
  console.log(`   Total Size: ${(totalSize / 1024).toFixed(2)} KB`);
  console.log(`   Total Semantic Blocks: ${totalBlocks}`);
  console.log(`   Average Compression: ${(totalTokens / totalSize).toFixed(4)} tokens/byte`);
  console.log(`   Storage Efficiency: ${(totalBlocks / totalSize * 1000).toFixed(2)} blocks/KB`);
  
  // Compare to hypothetical uncompressed JSON
  const estimatedJsonSize = totalTokens * 15; // Rough JSON overhead estimate
  const compressionSaved = ((estimatedJsonSize - totalSize) / estimatedJsonSize * 100);
  console.log(`   Estimated Compression vs JSON: ${compressionSaved.toFixed(1)}%`);
  console.log('');
  
  if (compressionSaved > 90) {
    console.log('🎉 AICF achieves >90% compression while maintaining semantic structure!');
  } else {
    console.log('📊 AICF provides significant compression with semantic benefits');
  }
}

if (require.main === module) {
  main();
}

module.exports = { analyzeFile, countTokens };

File: legacy-js/test-aicf-efficiency.py
#!/usr/bin/env python3

"""
AICF Token Efficiency Testing
Measures actual compression rates and semantic preservation
"""

import os
import re

def count_tokens(text):
    """Approximate GPT-style tokenization"""
    text = re.sub(r'\s+', ' ', text)
    return len([word for word in re.split(r'[\s\W]+', text) if word])

def analyze_file(file_path):
    """Analyze a single AICF file for efficiency metrics"""
    if not os.path.exists(file_path):
        print(f"❌ File not found: {file_path}")
        return None

    with open(file_path, 'r', encoding='utf-8') as f:
        content = f.read()
    
    lines = [line for line in content.split('\n') if line.strip()]
    tokens = count_tokens(content)
    size = len(content.encode('utf-8'))
    semantic_blocks = len(re.findall(r'@[A-Z_]+', content))

    return {
        'path': os.path.basename(file_path),
        'lines': len(lines),
        'tokens': tokens,
        'size': size,
        'compression_ratio': tokens / size if size > 0 else 0,
        'semantic_blocks': semantic_blocks
    }

def main():
    print('🧪 AICF Token Efficiency Analysis\n')
    
    aicf_dir = './.aicf'
    files = ['index.aicf', 'conversations.aicf', 'decisions.aicf', 'work-state.aicf', 'technical-context.aicf']
    
    total_tokens = 0
    total_size = 0
    total_blocks = 0
    
    for file in files:
        analysis = analyze_file(os.path.join(aicf_dir, file))
        if analysis:
            print(f"📊 {analysis['path']}:")
            print(f"   Lines: {analysis['lines']:,}")
            print(f"   Tokens: {analysis['tokens']:,}")
            print(f"   Size: {analysis['size']:,} bytes")
            print(f"   Semantic Blocks: {analysis['semantic_blocks']}")
            print(f"   Token/Byte Ratio: {analysis['compression_ratio']:.4f}")
            print('')
            
            total_tokens += analysis['tokens']
            total_size += analysis['size']
            total_blocks += analysis['semantic_blocks']
    
    if total_size > 0:
        print('📈 OVERALL STATISTICS:')
        print(f"   Total Tokens: {total_tokens:,}")
        print(f"   Total Size: {total_size / 1024:.2f} KB")
        print(f"   Total Semantic Blocks: {total_blocks}")
        print(f"   Average Compression: {total_tokens / total_size:.4f} tokens/byte")
        print(f"   Storage Efficiency: {total_blocks / total_size * 1000:.2f} blocks/KB")
        
        # Compare to hypothetical uncompressed JSON
        estimated_json_size = total_tokens * 15  # Rough JSON overhead estimate
        compression_saved = ((estimated_json_size - total_size) / estimated_json_size * 100)
        print(f"   Estimated Compression vs JSON: {compression_saved:.1f}%")
        print('')
        
        if compression_saved > 90:
            print('🎉 AICF achieves >90% compression while maintaining semantic structure!')
        else:
            print('📊 AICF provides significant compression with semantic benefits')
    else:
        print('❌ No valid files found for analysis')

if __name__ == "__main__":
    main()